src/network/networkd-link.c

```diff
@@ -2714,8 +2765,18 @@ int link_update(Link *link, sd_netlink_message *m) {
         }
 
         r = sd_netlink_message_read_strv(m, IFLA_PROP_LIST, IFLA_ALT_IFNAME, &s);
-        if (r >= 0)
+        if (r >= 0) {
+                char **n;
+
+                STRV_FOREACH(n, link->alternative_names)
+                        hashmap_remove_value(link->manager->links_by_name, *n, link);
+
+                r = link_update_by_name(link);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to add alternative name '%s' on network interface '%s', ignoring: %m", *n, link->ifname);
+
                 strv_free_and_replace(link->alternative_names, s);
+        }
```

Comment:
the main name of an iface can chance too, and needs to be covered. if you properly encapsulate the update/removal functions as suggested, then you will get this for free, as one will remove the whole shebang including main name, and the other will register the whole shebang

Reply:
I did not got this.

Reply:
the main name of the interface can change anytime too, and we should cover that here too and update the by-name hash tables, too